docs/design/pr-builder.rst

```diff
@@ -0,0 +1,143 @@
+Design of Pull Request Builder
+==============================
+
+Background
+----------
+
+This will focus on automatically building documentation for Pull Requests on Read the Docs projects.
+This is one of the most requested feature of Read the Docs.
+This document will serve as a design document for discussing how to implement this features.
+
+Scope
+-----
+
+.. contents::
+   :local:
+   :depth: 1
+   :backlinks: none
+   
+- Receiving ``pull_request`` webhook event from Github
+- Fetching data from pull requests.
+- Making Pull Requests work like temporary ``Version``
+- Excluding PR Versions from Elasticsearch Indexing
+- Creating PR Versions when a pull request is opened and Triggering a build
+- Triggering Builds on new commits on a PR
+- Status reporting to Github
+- Storing PR Version build Data
+- Excluding PR Versions from Search Engines
+- Serving PR Docs
+- Updating the Footer API
+- Adding Warning Banner to Docs
+
+Fetching Data from Pull Requests
+--------------------------------
+
+We already get Pull request events from Github webhooks.
+We can utilize that to fetch data from pull requests.
+when a ``pull_request`` event is triggered we can fetch the data of that pull request.
+We can fetch the pull request by doing something similar to travis-ci.
+ie: ``git fetch origin +refs/pull/<pr_number>/merge:``
+
+Modeling Pull Requests as a Type of Version
+-------------------------------------------
+
+Pull requests can be Treated as a Type of Temporary ``Version``.
+We might consider adding a ``VERSION_TYPES`` to the ``Version`` model.
```

Comment:
Thinking more about this, any reason why we want to merge the PR type with the "normal" ones?

Feels like more work trying to exclude all Pull Request versions, instead we could just have something like  `Project.temporal_versions`

I can't think of a case where we want this type listed in the same place with the normal PRs.

Reply:
This would be using the same Version model, just in a different column

Reply:
Then we would make the default `Version` manager exclude those versions, correct?

Reply:
Updated the PR with Versions managers (`internal`, `external`).